tools/test-locked-requirements

```diff
@@ -1,55 +1,38 @@
 #!/usr/bin/env python3
-import difflib
-import filecmp
+import glob
+import hashlib
 import os
-import subprocess
-import sys
-
-from typing import Text
 
 TOOLS_DIR = os.path.dirname(__file__)
 ZULIP_PATH = os.path.dirname(TOOLS_DIR)
-sys.path.append(ZULIP_PATH)
-from scripts.lib.zulip_tools import run
-
-PROD_LOCK_FILE = os.path.join(ZULIP_PATH, 'requirements', 'prod_lock.txt')
-DEV_LOCK_FILE = os.path.join(ZULIP_PATH, 'requirements', 'dev_lock.txt')
-TEST_PROD_LOCK_FILE = '/var/tmp/test_prod_lock.txt'
-TEST_DEV_LOCK_FILE = '/var/tmp/test_dev_lock.txt'
-
-def print_diff(path_file1, path_file2):
-    # type: (Text, Text) -> None
-    with open(path_file1) as file1:
-        with open(path_file2) as file2:
-            diff = difflib.unified_diff(
-                file1.readlines(),
-                file2.readlines(),
-                fromfile=path_file1,
-                tofile=path_file2,
-            )
-    for line in diff:
-        print(line)
+REQUIREMENTS_HASH_FILE = os.path.join(ZULIP_PATH, 'requirements', 'requirements_hash')
+
+def get_requirements_hash():
+    # type: () -> List[str]
+    hash_values = []
+    for file_path in glob.glob(os.path.join(ZULIP_PATH, "requirements", "*.txt")):
+        with open(file_path) as fp:
+            sha1 = hashlib.sha1(fp.read().encode("utf-8"))
+            hash_values.append(sha1.hexdigest() + '\n')
+    sha1 = hashlib.sha1()
+    for hash_val in sorted(hash_values):
+        sha1.update(hash_val.encode("utf-8"))
+    return sha1.hexdigest()
+
+def locked_requirements_are_stale():
+    # type: () -> bool
+    with open(REQUIREMENTS_HASH_FILE) as fp:
+        for line in fp.readlines():
+            if not line.startswith('#'):
+                prev_hash = line.strip()
+    curr_hash = get_requirements_hash()
+    if curr_hash != prev_hash:
+        return True
+    return False
 
 def main():
     # type: () -> None
-    # `pip-compile` tries to avoid unnecessarily updating recursive dependencies
-    # if lock files are present already. If we don't copy these files to the tmp
-    # dir then recursive dependencies will get updated to their latest version
-    # without any change in the input requirements file and the test will not pass.
-    run(['cp', PROD_LOCK_FILE, TEST_PROD_LOCK_FILE])
-    run(['cp', DEV_LOCK_FILE, TEST_DEV_LOCK_FILE])
-    subprocess.check_call([os.path.join(TOOLS_DIR, 'update-locked-requirements'),
-                           '--prod', TEST_PROD_LOCK_FILE, '--dev', TEST_DEV_LOCK_FILE,
-                           ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```

Comment:
This pre-copy behavior was important for avoiding `pip-compile` trying to update the versions of recursive dependencies to the latest version every time while avoiding patching files locally; why did you remove it?

Reply:
I have actually tried to change the implementation here. Now I am not running the pip-compile at all in the test. Instead of that I am actually trying the verify the state of the requirements files using a hash which update-locked-requirements produces. I recalculate the hash in the test and match it against the stored version. If it matches then it confirms that user actually ran the tool to update the locked requirements files.

Reply:
Hmm, I don't think that's going to work in Travis CI, unless the model is that we check in the hash file.  Ohh, I see, that's exactly what you're doing.  OK.

Reply:
The thing I don't like about this model is that we end up making any two PRs that change dependencies merge conflict with each other (since they'll collide on that has file).  

Whereas with the old model for how this test worked, it two totally different parts of the `requirements.txt` files (and locks) are changed in a way where `git rebase` merges correctly, it'll just work.

Reply:
I have actually left a comment in the hash file regarding this, the simplest way to deal with this situation is to simply delete the hash file and regenerate it using the update-locked-requirements tool.
I was unable to come up with a solution that could avoid running the expensive old test using some caching in travis.

Reply:
I think there is a solution to that where we just store in Travis CI the pairs of 
(requirements_hash, locked_requirements_hash) that are known-valid?

@hackerkid how annoying do you think it would be to have to remove-regenerate this file on every change?